src/app/rosetta/lib/rosetta.ml

```diff
@@ -2,49 +2,51 @@ open Core_kernel
 open Async
 open Rosetta_lib
 
-let router ~graphql_uri ~db ~logger route body =
-  try
-    match route with
-    | "network" :: tl ->
-        Network.router tl body ~db ~graphql_uri ~logger
-    | "account" :: tl ->
-        Account.router tl body ~db ~graphql_uri ~logger
-    | "mempool" :: tl ->
-        Mempool.router tl body ~db ~graphql_uri ~logger
-    | "block" :: tl ->
-        Block.router tl body ~db ~graphql_uri ~logger
-    | "construction" :: tl ->
-        Construction.router tl body ~graphql_uri ~logger
-    | _ ->
-        Deferred.return (Error `Page_not_found)
-  with exn -> Deferred.return (Error (`Exception exn))
-
 let server_handler ~pool ~graphql_uri ~logger ~body _sock req =
+  let with_db f =
+    let open Deferred.Result.Let_syntax in
+    let%bind pool = Lazy.force pool in
+    Caqti_async.Pool.use (fun db -> f ~db) pool
+    |> Deferred.Result.map_error ~f:(function
+         | `App e ->
+             `App e
+         | `Page_not_found ->
+             `Page_not_found
+         | `Exception exn ->
+             `Exception exn
+         | `Connect_failed _e ->
+             `App (Errors.create (`Sql "Connect failed"))
+         | `Connect_rejected _e ->
+             `App (Errors.create (`Sql "Connect rejected"))
+         | `Post_connect _e ->
+             `App (Errors.create (`Sql "Post connect error")) )
+  in
+  let router route body =
```

Comment:
This doesn't seem to depend on any of the function arguments here. Can you move it back out?

Reply:
It depends on `with_db` but I'll move `router` back out and move `with_db` to `router`, that's nicer